### PR TITLE
[MM-43149] Add first part of Elrond code

### PR DESCRIPTION
**Pull request by stylianosrigas**
#### Summary
  
  - Add first part of Elrond code
  - Database handling
  - Basic ring functionality
  - Cmd and API support
  
  The code is work in progress and its missing the main ring functionality.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	$(GO) vet ./...
+	@echo Govet success
+
+.PHONY: unittest
+unittest:
+	$(GO) test ./... -v -covermode=count -coverprofile=coverage.out
+
+# Build for distribution
+.PHONY: build
+build:
```
Comment (stafot) [inline Makefile:40]: vvvv minor: suggesting to support multiarch builds

File: build/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+
+FROM golang:1.17.3-alpine3.14 AS builder
```
Comment (stafot) [inline build/Dockerfile:2]: if there are no strong objections let's use images without patch versions if available to get the latest patch version on the build automatically

Comment (stylianosrigas) [reply build/Dockerfile:2]: I haven't focused on dockerfile at all yet. I will work on that tomorrow.

File: build/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+COPY . .
+RUN CGO_ENABLED=0 make build
+
+#####
+
+FROM alpine:3.14
+
+COPY --from=builder /go/src/fleet-controller/fleet-controller /fleet-controller
+
+ENTRYPOINT [ "/fleet-controller" ]
```
Comment (stafot) [inline build/Dockerfile:21]: @stylianosrigas it is obvious that you didn't touch it yet

File: cmd/elrond/ring.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+			return errors.Wrap(err, "provided server address not a valid address")
+		}
+
+		client := model.NewClient(serverAddress)
+
+		ringID, _ := command.Flags().GetString("ring")
+
+		err := client.DeleteRing(ringID)
+		if err != nil {
+			return errors.Wrap(err, "failed to delete ring")
```
Comment (stafot) [inline cmd/elrond/ring.go:246]: Could you return the `ringID` in all the relevant errors please?

File: build/Dockerfile (hunk 2 above)
Comment (stylianosrigas) [reply build/Dockerfile]: Dockerfile improvements pushed

File: Makefile (hunk 1 above)
Comment (stylianosrigas) [reply Makefile:65]: done thanks :)

File: cmd/elrond/ring.go (hunk 4 above)
Comment (stylianosrigas) [reply cmd/elrond/ring.go]: Done, thanks

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	bash ./scripts/push-image.sh
+
+.PHONY: install
+install: build
+	go install ./...
+
+.PHONY: release
+release:
+	@echo Cut a release
+	bash ./scripts/release.sh
```
Comment (stafot) [inline Makefile:108]: Since you added this you need to add the relevant script and the relevant CI logic.

Comment (stylianosrigas) [reply Makefile:108]: Yeap already did, just forgot to push thanks ;)

File: scripts/release.sh [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+FUTURE_RELEASE_SHA=$(hub rev-parse HEAD)
+LATEST_RELEASE=$(hub release | head -n 1)
+# The following fixes the script to work in the case that there isn't a previous release in place.
+if [[ "$LATEST_RELEASE" == "" ]]; then
+  LATEST_RELEASE=$(git log --pretty=%H | tail -n 1)
+fi
+LATEST_RELEASE_SHA=$(hub rev-parse "${LATEST_RELEASE}")
+LATEST_RELEASE_NEXT_COMMIT_SHA=$(hub log "${LATEST_RELEASE}"..HEAD --oneline --pretty=%H | tail -n1)
+mkdir -p ./build/_output/docs/
+release-notes --org mattermost --repo elrond --start-sha "${LATEST_RELEASE_NEXT_COMMIT_SHA}" --end-sha "${FUTURE_RELEASE_SHA}"  --output ./build/_output/docs/relnote.md --required-author "" --branch master
```
Comment (stafot) [inline scripts/release.sh:15]:
```suggestion
release-notes --org mattermost --repo elrond --start-sha "${LATEST_RELEASE_NEXT_COMMIT_SHA}" --end-sha "${FUTURE_RELEASE_SHA}"  --output ./build/_output/docs/relnote.md --required-author "" --branch main
```

File: scripts/push-image-pr.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+set -e
+set -u
+
+env >> BASH_ENV
+cat BASH_ENV | while read line; do
+	export $line
+done
+
+export TAG="${CIRCLE_SHA1:0:7}"
```
Comment (stafot) [inline scripts/push-image-pr.sh:10]: Assumes usage of `CircleCI` need to be changed accordingly with the value of the preffered CI or make it variable to be passed at runtime.

File: scripts/push-image.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+set -e
+set -u
+
+env >> BASH_ENV
+cat BASH_ENV | while read line; do
+	export $line
+done
+if [ -z "${CIRCLE_TAG:-}" ]; then
```
Comment (stafot) [inline scripts/push-image.sh:9]: Assumes usage of `CircleCI` need to be changed accordingly with the value of the preffered CI or make it variable to be passed at runtime.

**Review by stafot — APPROVED** [review]
LGTM. 
Minor suggestions for scripts.

**Review by angeloskyratzakos — APPROVED** [review]

File: internal/store/migrations.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+			return err
+		}
+
+		if _, err = e.Exec(`
+			CREATE TABLE Ring (
+				ID CHAR(26) PRIMARY KEY,
+				State TEXT NOT NULL,
+				Name TEXT NOT NULL,
+				Priority INT NOT NULL,
+				InstallationGroup STRING NULL,
```
Comment (Szymongib) [inline internal/store/migrations.go:40]: I think Postgres will not accept such type.

```suggestion
				InstallationGroup TEXT NULL,
```

File: internal/store/ring.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	defer tx.RollbackUnlessCommitted()
+
+	if err = sqlStore.createRing(tx, ring); err != nil {
+		return errors.Wrap(err, "failed to create ring")
+	}
+
+	if err = tx.Commit(); err != nil {
+		return errors.Wrap(err, "failed to commit the transaction")
+	}
```
Comment (Szymongib) [inline internal/store/ring.go:120]: When running only one DB operation you do not need to start transaction separately, so you could just do `sqlStore.createRing(sqlStore.db, ring)` unless you want to add more things there in the future.

File: model/ring_request.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+// CreateRingRequest specifies the parameters for a new ring.
+type CreateRingRequest struct {
+	Name              string `json:"name,omitempty"`
+	Priority          int    `json:"priority,omitempty"`
+	InstallationGroup string `json:"installationGroup,omitempty"`
+	SoakTime          int    `json:"soakTime,omitempty"`
+	Image             string `json:"image,omitempty"`
+	Version           string `json:"version,omitempty"`
+	APISecurityLock   bool   `json:"api-security-lock,omitempty"`
```
Comment (Szymongib) [inline model/ring_request.go:24]: I would suggest keeping it consistent if we use camel case for others.
```suggestion
	APISecurityLock   bool   `json:"apiSecurityLock,omitempty"`
```

**Review by Szymongib — COMMENTED** [review]
Sorry for the late review. Did not manage to go through the whole thing and I know you want to move forward on that so do not consider it blocking. Left few small suggestions.

File: model/ring_request.go (hunk 11 above)
Comment (stylianosrigas) [reply model/ring_request.go]: Thanks for the suggestion, fixed

File: internal/store/migrations.go (hunk 9 above)
Comment (stylianosrigas) [reply internal/store/migrations.go]: Thanks for the suggestion, fixed

File: scripts/release.sh (hunk 6 above)
Comment (stylianosrigas) [reply scripts/release.sh]: Thanks for the suggestion, fixed

**Comment by stylianosrigas** [comment]
I will merge this for now, as its needed to mover forward and will fit the pending nit comments in the next one.

**Pull request merged**